src/js/jsx/sections/layers/LayerGroup.jsx

```diff
@@ -0,0 +1,137 @@
+
+    var LayerGroup = React.createClass({
+        mixins: [FluxMixin],
+        
+        /**
+         * True if the LayerGroup's components are rendered. Used to avoid
+         * rendering layer group until it is first visible.
+         * 
+         * @type {Boolean}
+         */
+        isRendered: false,
+        
+        propTypes: {
+            /**
+             * List of Layer nodes in the same layer group.
+             * @type {Immutable.List.<LayerNode>}
+             */
+            layerNodes: React.PropTypes.object.isRequired,
+
+            /**
+             * The parent layer of the layer nodes.
+             * @type {Layer}
+             */
+            parentLayer: React.PropTypes.object,
+            
+            /**
+             * @type {Document}
+             */
+            document: React.PropTypes.object.isRequired,
+            
+            /**
+             * @type {Boolean}
+             */
+            disabled: React.PropTypes.bool
+        },
+
+        render: function () {
+            var parentLayer = this.props.parentLayer,
+                isRoot = !parentLayer;
+
+            // Do not render layer group of collapsed layer to reduce document load time.
+            if (!isRoot && !parentLayer.expanded && !this.isRendered) {
+                return false;
+            } else {
+                this.isRendered = true;
+            }
+
+            var layerFaces = this.props.layerNodes.reduce(function (results, layerNode) {
+                var layer = this.props.document.layers.byID(layerNode.id);
+
+                if (!layer.isGroupEnd) {
+                    var layerGroup = layerNode.children && (
+                        <LayerGroup
+                            parentLayer={layer}
+                            layerNodes={layerNode.children}
+                            document={this.props.document}
+                            disabled={this.props.disabled}/>
+                    );
+
+                    var className = classnames({
+                        "layer-group": layerGroup,
+                        "layer-group__selected": layerGroup && layer.selected,
+                        "layer-group__collapsed": layerGroup && !layer.expanded,
+                        "layer-group__not-visible": layerGroup && !layer.visible
+                    });
+
+                    results.push(
+                        <li key={layer.key} className={className}>
+                            <LayerFace
+                                key={layer.key}
+                                disabled={this.props.disabled}
+                                document={this.props.document}
+                                layer={layer}
+                                childNodes={layerNode.children}/>
+                            {layerGroup}
+                        </li>
+                    );
+                }
+                
+                return results;
+            }.bind(this), []);
```

Comment:
I believe `Immutable.Iterable.prototype.reduce` supports a third `receiver` argument. We prefer that, when available, over explicit calls to `.bind`.